File: leetcode/golang/medium/Problem36.go
package main

import "fmt"

// Hint: record number for rows, columns, squares separately
func isValidSudoku(board [][]byte) bool {
	rows := [9][9]bool{}
	cols := [9][9]bool{}
	squares := [3][3][9]bool{}

	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			if board[i][j] == '.' {
				continue
			}
			num := int(board[i][j]) - 49 // ascii value of '1' is 49
			if rows[i][num] || cols[j][num] || squares[i/3][j/3][num] {
				return false
			}

			rows[i][num] = true
			cols[j][num] = true
			squares[i/3][j/3][num] = true
		}
	}
	return true
}

// Leetcode 36
// Link: https://leetcode.com/problems/valid-sudoku/description/
func main() {
	sudoku1 := [][]byte{
		{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
		{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
		{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
		{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
		{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
		{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
		{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
		{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
		{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	}
	fmt.Println(isValidSudoku(sudoku1))

	sudoku2 := [][]byte{
		{'8', '3', '.', '.', '7', '.', '.', '.', '.'},
		{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
		{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
		{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
		{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
		{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
		{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
		{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
		{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	}
	fmt.Println(isValidSudoku(sudoku2))
}

File: leetcode/python/medium/Problem36.py
# Questions to ask:
# 1. What is the time complexity? O(1) because only 81 cells
# 2. What is the space complexity? O(1) because only 81 cells
class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        cols = [{str(i+1): False for i in range(9)} for _ in range(9)]
        rows = [{str(i+1): False for i in range(9)} for _ in range(9)]
        boxes = [{str(i+1): False for i in range(9)} for _ in range(9)]
        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num == ".":
                    continue
                box = (i // 3) * 3 + j // 3
                if cols[j][num] or rows[i][num] or boxes[box][num]:
                    return False
                cols[j][num] = True
                rows[i][num] = True
                boxes[box][num] = True
        return True


# Problem 36
# Link: https://leetcode.com/problems/valid-sudoku/description/
if __name__ == '__main__':
    s = Solution()
    cases = [
        (
            [
                ["5", "3", ".", ".", "7", ".", ".", ".", "."],
                ["6", ".", ".", "1", "9", "5", ".", ".", "."],
                [".", "9", "8", ".", ".", ".", ".", "6", "."],
                ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
                ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
                ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
                [".", "6", ".", ".", ".", ".", "2", "8", "."],
                [".", ".", ".", "4", "1", "9", ".", ".", "5"],
                [".", ".", ".", ".", "8", ".", ".", "7", "9"]
            ],
            True
        )
    ]
    for board, expected in cases:
        assert s.isValidSudoku(board) == expected
